### CS559 - Homework #1

**Author**: Sid Bhatia

**Date**: September 7th, 2024

**Pledge**: I pledge my honor that I have abided by the Stevens Honor System.

**Professor**: Dr. In Suk Jang

#### 1. Linear Regression Modification [40 pts]

In the lecture, the linear regression algorithm, `Linear_Regression(learning_rate, iteration)`, was implemented to solve a weight vector in a gradient descent fashion. In this assignment, we will modify the linear regression that optimizes the weights via stochastic gradient descent and predict the house price of the unit area.

a. Load the provided data set as follows.

```python
import pandas as pd
df = pd.read_csv('./Realestate.csv')
```

In [27]:
import pandas as pd
df = pd.read_csv('./Realestate.csv')

b. [5pts] Standardize the data using each feature’s mean $\mu$ and standard deviation $\sigma_i$ as shown below:

$$
x_{i, \text{std}} = \frac{x_i - \mu_i}{\sigma_i}
$$

where $i$ is the feature index. Standardize the target variable as well. The standardized data set will be used for the following questions.

In [28]:
# View overhead of DataFrame
print(df.head())

# Retrieve column names and parse as a list
col_names = df.columns.values.tolist()

# View column names
print(col_names)

# Retrieve target variable name
target = col_names[-1]

print(target)

   No  X1 transaction date  X2 house age  \
0   1             2012.917          32.0   
1   2             2012.917          19.5   
2   3             2013.583          13.3   
3   4             2013.500          13.3   
4   5             2012.833           5.0   

   X3 distance to the nearest MRT station  X4 number of convenience stores  \
0                                84.87882                               10   
1                               306.59470                                9   
2                               561.98450                                5   
3                               561.98450                                5   
4                               390.56840                                5   

   X5 latitude  X6 longitude  Y house price of unit area  
0     24.98298     121.54024                        37.9  
1     24.98034     121.53951                        42.2  
2     24.98746     121.54391                        47.3  
3     24.98746     121.54391  

In [29]:
# Separate features and target
features = df.drop(columns=['Y house price of unit area'])
target = df['Y house price of unit area']

# Standardized features and target
features_stand = (features - features.mean()) / features.std()

target_stand = (target - target.mean()) / target.std()

# print(features_stand.head())

# df_merged = features_stand.append(target_stand) DEPRECIATED

df_stand = pd.concat([features_stand, target_stand.rename('Y house price of unit area')], axis = 1)

print(df_stand.head())
print(target_stand.head())

         No  X1 transaction date  X2 house age  \
0 -1.725784            -0.822688      1.254111   
1 -1.717427            -0.822688      0.156896   
2 -1.709069             1.539289     -0.387322   
3 -1.700712             1.244928     -0.387322   
4 -1.692355            -1.120595     -1.115873   

   X3 distance to the nearest MRT station  X4 number of convenience stores  \
0                               -0.791537                         2.004982   
1                               -0.615866                         1.665488   
2                               -0.413515                         0.307513   
3                               -0.413515                         0.307513   
4                               -0.549332                         0.307513   

   X5 latitude  X6 longitude  Y house price of unit area  
0     1.124070      0.448220                   -0.005894  
1     0.911342      0.400654                    0.310132  
2     1.485063      0.687352                    0.684

c. [5pts] Using `Scikit-Learn Linear Regression`, predict the target and report the Root Mean Square Error Value (RMSE).

In [30]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

import numpy as np

# Define feature and target variable
X = df_stand[['X2 house age', 'X3 distance to the nearest MRT station']]
y = df_stand[['Y house price of unit area']]

# Create Linear Regression model and fit it
model = LinearRegression()

# Fit Linear Regression model
model.fit(X, y)

# Establish y_hat to predict on training set
y_hat = model.predict(X)

# Retrieve mean squared error
mse = mean_squared_error(y, y_hat)

# Calculate root mean squared error
rmse = np.sqrt(mse)

print(f"Root Mean Square Error (RMSE): {rmse:.4f}")

Root Mean Square Error (RMSE): 0.7125


d. Using the `NumPy Linear_Regression()` implemented in the lecture, predict the target and report the RMSE value. Tune learning rate $\eta$ and the iteration number so the RMSE value is not more than $10\%$ of the RMSE value from the Scikit-learn Linear Regression result. Report the final RMSE and $\eta$ values.

In [31]:
class Linear_regression():
    
    def __init__(self, learning_rate, iterations):
        self.learning_rate = learning_rate
        self.iterations = iterations
    
    # Model Training Function
    def fit(self, X, Y):
        
        # Number of training example, Number of features
        self.m, self.n = X.shape 
        
        # Weight initialization
        
        self.W = np.zeros(self.n)
        self.b = 0
        self.X = X
        self.Y = Y
        
        # Gradient Descent Learning    
        for i in range(self.iterations):
            self.update_weights()
        return self
    
    # Function to update weights in Gradient Descent
    def update_weights(self):
        Y_pred = self.predict(self.X)
        
        # Gradient calculations
        dW = - (2 * (self.X.T).dot(self.Y - Y_pred)) / self.m
        db = - 2 * np.sum(self.Y - Y_pred) / self.m
        
        # Weight updates
        self.W = self.W - self.learning_rate * dW
        self.b = self.b - self.learning_rate * db
        
        return self
    # Hypthetical Function
    def predict(self, X):
        
        return X.dot(self.W) + self.b

In [32]:
# Arrange variables to fit Linear_regression()
x1 = np.array(df_stand['X2 house age'])
x2 = np.array(df_stand['X3 distance to the nearest MRT station'])

y = np.array(target_stand)

X = np.column_stack([x1, x2])

In [33]:
model = Linear_regression(learning_rate = 0.005, iterations=10000)

# Fit the model
model.fit(X, y)

# Make predictions
h_hat = model.predict(X)

# Print the learned weights and bias
print(f'W = {model.W}')
print(f'b = {model.b}')

# Calculate RMSE
mse = mean_squared_error(y, h_hat)
rmse = np.sqrt(mse)
print(f'Final RMSE = {rmse:.4f}')

W = [-0.19343469 -0.66865666]
b = -6.972415130496186e-18
Final RMSE = 0.7125


e. [20pts] Modify the `Linear_regression()` algorithm that estimates the weights in a **stochastic gradient descent** manner. Report the RMSE value. Start with the $\eta$ found from (d) and tune the parameters until the RMSE value is within $10\%$ of the RMSE found from (c).

In [34]:
class Linear_regression():
    
    def __init__(self, learning_rate, iterations):
        self.learning_rate = learning_rate
        self.iterations = iterations
    
    # Model Training Function
    def fit(self, X, Y):
        
        # Number of training example, Number of features
        self.m, self.n = X.shape 
        
        # Weight initialization
        
        self.W = np.zeros(self.n)
        self.b = 0
        self.X = X
        self.Y = Y
        
        # Stochastic Gradient Descent    
        for i in range(self.iterations):

            # Shuffle the data.
            indices = np.random.permutation(self.m)
            X_shuffled = self.X[indices]
            Y_shuffled = self.Y[indices]

            # Iterate through each sample.
            for j in range(self.m):
                self.update_weights(X_shuffled[j], Y_shuffled[j])
        return self
    
    # Function to update weights in Gradient Descent
    def update_weights(self, X_sample, Y_sample):
        Y_pred = self.predict(X_sample)
        
        # Calculate gradients
        dW = -2 * (X_sample.T * (Y_sample - Y_pred))  # Gradient of weight
        db = -2 * (Y_sample - Y_pred)  # Gradient of bias
        
        # Update weights and bias
        self.W = self.W - self.learning_rate * dW
        self.b = self.b - self.learning_rate * db
        
        return self
    
    # Hypthetical Function
    def predict(self, X):
        return X.dot(self.W) + self.b

In [35]:
# Arrange variables to fit Linear_regression().
x1 = np.array(df_stand['X2 house age'])
x2 = np.array(df_stand['X3 distance to the nearest MRT station'])

y = np.array(target_stand)

X = np.column_stack([x1, x2])

In [36]:
model = Linear_regression(learning_rate = 0.002, iterations=10000)

# Fit the model
model.fit(X, y)

# Make predictions
h_hat = model.predict(X)

# Print the learned weights and bias
print(f'W = {model.W}')
print(f'b = {model.b}')

# Calculate RMSE
mse = mean_squared_error(y, h_hat)
rmse = np.sqrt(mse)
print(f'Final RMSE = {rmse:.4f}')

W = [-0.20245669 -0.63883477]
b = -0.019426865108235394
Final RMSE = 0.7134


#### 2. Bayesian Theorem [15pts]

In [37]:
import pandas as pd
df = pd.read_csv('./data.csv')

print(df.head())

   X  Y
0  1  1
1  0  0
2  0  0
3  1  0
4  1  0


b. [5pts] Calculate a prior probability of $x$ and $y$, $p(x = 0), p(x = 1), p(y = 0)$ and $p(y = 1)$.

In [38]:
data = {'X': [1, 0, 0, 1, 1],
        'Y': [1, 0, 0, 0, 0]}
df = pd.DataFrame(data)

print(df.head())

   X  Y
0  1  1
1  0  0
2  0  0
3  1  0
4  1  0


In [39]:
# Calculate prior probabilities
p_x_0 = len(df[df['X'] == 0]) / len(df)

print(p_x_0)

p_x_1 = len(df[df['X'] == 1]) / len(df)

print(p_x_1)

p_y_0 = len(df[df['Y'] == 0]) / len(df)

print(p_y_0)

p_y_1 = len(df[df['Y'] == 1]) / len(df)

print(p_y_1)

0.4
0.6
0.8
0.2


c. [5pts] Calculate the likelihood of $x$, $p(x = 0 \mid y = 0), p(x = 1 \mid y = 0), p(x = 0 \mid y = 1)$, and $p(x = 1 \mid y = 1)$.

In [40]:
# Calculate the likelihoods
p_x_0_given_y_0 = len(df[(df['X'] == 0) & (df['Y'] == 0)]) / len(df[df['Y'] == 0])

print(p_x_0_given_y_0)

p_x_1_given_y_0 = len(df[(df['X'] == 1) & (df['Y'] == 0)]) / len(df[df['Y'] == 0])

print(p_x_1_given_y_0)

p_x_0_given_y_1 = len(df[(df['X'] == 0) & (df['Y'] == 1)]) / len(df[df['Y'] == 1])

print(p_x_0_given_y_1)

p_x_1_given_y_1 = len(df[(df['X'] == 1) & (df['Y'] == 1)]) / len(df[df['Y'] == 1])

print(p_x_1_given_y_1)

0.5
0.5
0.0
1.0


d. [5pts] Calculate the posterior probabilities, $p(y = 1 \mid x = 0), p(y = 0 \mid x = 0), p(y = 1 \mid x = 1)$, and $p(y = 0 \mid x = 1)$.

In [46]:
# Calculating posteriors using Bayes' Theorem
p_y_1_given_x_0 = (p_x_0_given_y_1 * p_y_1) / p_x_0

print(p_y_1_given_x_0)

p_y_0_given_x_0 = (p_x_0_given_y_0 * p_y_0) / p_x_0

print(p_y_0_given_x_0)

p_y_1_given_x_1 = (p_x_1_given_y_1 * p_y_1) / p_x_1

print(p_y_1_given_x_1)

p_y_0_given_x_1 = (p_x_1_given_y_0 * p_y_0) / p_x_1

print(p_y_0_given_x_1)

0.0
1.0
0.33333333333333337
0.6666666666666667


#### 3. LDA [45pts]

The binary LDA classification of $\bf{X}^{2 \times 2}$ was solved by finding the direction of the projection hyperplane $(\mathbf{w \propto S^{-1}_w(m_2 - m_1)})$ in the lecture. In this problem, the binary LDA classification model will be trained and the train model will generalize the test data. Please display the results for full credit consideration.

Generate the data set and split it into train and test sets.

```python
from sklearn import datasets
import numpy as np

X, y = datasets.make_blobs(n_samples = 100, n_features = 4, centers = 2, cluster_std = 1.5, random_state = 123)

y = np.array([-1 if i == 0 else 1 for i in y])

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 42)
```

In [47]:
from sklearn import datasets
import numpy as np

X, y = datasets.make_blobs(n_samples = 100, n_features = 4, centers = 2, cluster_std = 1.5, random_state = 123)

y = np.array([-1 if i == 0 else 1 for i in y])

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 42)

a. [10pts] Write a method that computes the between-class scatter matrix $\bf{S}_B$. Compute $\bf{S}_B$ using a train data.

In [53]:
# Method to compute the between-class scatter matrix S_B
def compute_S_B(X_train, y_train):
    # Compute the class means
    class_labels = np.unique(y_train)
    mean_vectors = [np.mean(X_train[y_train == label], axis=0) for label in class_labels]
    
    # Overall mean
    overall_mean = np.mean(X_train, axis=0)
    
    # Compute between-class scatter matrix S_B
    S_B = np.zeros((X_train.shape[1], X_train.shape[1]))
    for i, mean_vec in enumerate(mean_vectors):
        n = X_train[y_train == class_labels[i]].shape[0]  # Number of samples in class
        mean_diff = (mean_vec - overall_mean).reshape(-1, 1)
        S_B += n * (mean_diff @ mean_diff.T)  # Accumulate the scatter

    return S_B

# Compute the between-class scatter matrix S_B
S_B = compute_S_B(X_train, y_train)

print(S_B)

[[ 1.42177386e+00 -1.48370842e+01 -7.83733634e+01 -1.38981884e+01]
 [-1.48370842e+01  1.54834094e+02  8.17874222e+02  1.45036139e+02]
 [-7.83733634e+01  8.17874222e+02  4.32022577e+03  7.66118860e+02]
 [-1.38981884e+01  1.45036139e+02  7.66118860e+02  1.35858203e+02]]


b. [10pts] Write a method that computes and returns the within-class scatter of each class, the total-within class scatter ($\bf{S}_w$), and its inverse ($\bf{S}^{-1}_w$). Calculate $\bf{S}_w$ and $S^{-1}_w$ using the train data.

In [58]:
# Method to compute the within-class scatter matrix S_w and its inverse S_w_inv
def compute_S_w(X_train, y_train):
    # Compute the class means
    class_labels = np.unique(y_train)
    
    # Initialize the within-class scatter matrix S_w
    S_w = np.zeros((X_train.shape[1], X_train.shape[1]))
    
    # Compute within-class scatter for each class
    for label in class_labels:
        class_scatter = np.cov(X_train[y_train == label], rowvar=False) * (len(X_train[y_train == label]) - 1)
        S_w += class_scatter  # Accumulate the scatter for each class

    # Calculate the inverse of the within-class scatter matrix
    S_w_inv = np.linalg.inv(S_w)
    
    return S_w, S_w_inv

# Compute the within-class scatter matrix S_w and its inverse S_w_inv using the training data
S_w, S_w_inv = compute_S_w(X_train, y_train)

print(S_w)
print("\n")
print(S_w_inv)

[[189.11331817   6.52860091 -16.42044417   8.18216257]
 [  6.52860091 204.02423606  -2.00586952   5.86496479]
 [-16.42044417  -2.00586952 154.61882164   2.54574474]
 [  8.18216257   5.86496479   2.54574474 147.72968011]]


[[ 5.35588997e-03 -1.57131974e-04  5.71698050e-04 -3.00255075e-04]
 [-1.57131974e-04  4.91228014e-03  5.01215342e-05 -1.87181528e-04]
 [ 5.71698050e-04  5.01215342e-05  6.53128940e-03 -1.46204087e-04]
 [-3.00255075e-04 -1.87181528e-04 -1.46204087e-04  6.79570108e-03]]


c. [10pts] `NumPy.linalg.eig()` estimates eigenvalues $\lambda$ and their corresponding eigenvectors $\bf{u}$. Using `NumPy.linalg.eig()`,  find the direction of all projection planes, and transform the train data to each projection plane. Determine the most ideal $\lambda$ and $\bf{u}$ for the binary classification. Explain why.

In [61]:
# Compute the matrix product of S_w_inv and S_B
matrix_product = S_w_inv @ S_B

# Compute eigenvalues and eigenvectors using numpy's linalg.eig()
eigenvalues, eigenvectors = np.linalg.eig(matrix_product)

# Display eigenvalues and eigenvectors
eigenvalues, eigenvectors

(array([ 0.00000000e+00,  2.96351674e+01, -3.58042237e-16,  1.60188508e-16]),
 array([[-0.99984555, -0.05874635,  0.1560219 , -0.17561664],
        [-0.00250869, -0.14249608,  0.85218794, -0.41879989],
        [-0.01718134, -0.97591602, -0.2365057 , -0.08123579],
        [-0.00271783, -0.15437502,  0.439884  ,  0.88722387]]))

In [60]:
projection_vector = eigenvectors[:, np.argmax(eigenvalues)]
X_train_transformed = X_train @ projection_vector

print("Transformed training data:")
print(X_train_transformed)

Transformed training data:
[-10.49344401  -8.23403937   6.23778243   5.08140446   6.72768436
 -10.93272746   6.64475407 -10.29143476   6.90063926   7.48667905
  -9.20802779   4.20466442   5.59896819  -7.66837953   3.54573059
 -10.25743448   5.49893221   4.2407423    5.27944813   5.34346907
 -11.11043834   4.91767919   6.66868634   6.90746514  -9.60594471
  -7.52798134  -7.2455161   -9.15383918   4.6228364    8.92768934
   1.94799768  -8.9249959  -12.15341543  -9.43776759   3.08870353
  -9.56730677   3.61019215   4.89304479  -8.39122099   3.51991109
  -9.41591469  -8.87569777 -11.22517511   4.29068917   3.95598145
  -8.67028721 -10.60203636 -11.52335963   5.18338783   7.30538879
 -10.33219259   5.94491653   6.18164718   5.80082771   7.14751497
 -11.99170186  -9.88351589 -10.12535624   6.28384107  -8.9039398
 -12.81448959 -10.37107446 -10.56376386   4.60320796  -9.58593198
   5.26765969   3.30152767  -7.96083662   7.36595707   5.11734733
   4.96162237 -11.23722769  -9.08314793   2.781980

The eigenvalue $\lambda = 29.64$ is the largest positive eigenvalue, which results in the largest variance between two classes in the projected direction. Therefore, it's the best for binary classification.

The corresponding eigenvector $u = [-0.0587, -0.1425, -0.9759, -0.1544]$ represents the direction of the projection plane which maximizes the separation between two classes.

d. [5pts] Using a choice of $\bf{u}$ from 3(c), predict the class of train data and calculate the accuracy score using the `from sklearn.metrics import accuracy_score`.

In [64]:
from sklearn.metrics import accuracy_score

# Choosing the eigenvector corresponding to the largest eigenvalue
best_u = eigenvectors[:, np.argmax(eigenvalues)]

# Project the training data onto this eigenvector
X_train_projection = X_train @ best_u

# Predict the classes: if projection > 0, classify as 1; otherwise, classify as -1
y_train_pred = np.where(X_train_projection > 0, 1, -1)

# Calculate the accuracy score
train_accuracy = accuracy_score(y_train, y_train_pred)
train_accuracy

0.0

e. [5pts] Predict the class of train data using other $\bf{u}$ vectors in 3(c) and calculate the accuracy. Based on the results of 3(c) and 3(d), explain if your choice of $\bf{u}$ is the correct decision.

In [65]:
# Collect the eigenvectors
other_us = [eigenvectors[:, i] for i in range(len(eigenvalues)) if i != np.argmax(eigenvalues)]

# Dictionary to store accuracy scores for each eigenvector
accuracy_scores = {}

# Loop through each of the other eigenvectors and calculate accuracy
for i, u in enumerate(other_us):
    X_train_projection_other = X_train @ u 
    threshold_other = np.mean(X_train_projection_other)
    y_train_pred_other = np.where(X_train_projection_other > threshold_other, 1, -1)
    accuracy = accuracy_score(y_train, y_train_pred_other)
    accuracy_scores[f'u_{i+1}'] = accuracy

accuracy_scores

{'u_1': 0.525, 'u_2': 0.475, 'u_3': 0.5375}

The largest eigenvalue chosen in 3(d) resulted in an accuracy score of $0\%$ which indicates while the corresponding eigenvector corresponds to the best direction of separating the classes based on variance, the overlap between the projections of the classes prevents effective classification with a basic threshold.

Therefore, the eigenvector corresponding to $\bf{u}_3$ turns out to be the best for classification based on the accuracy score.